### PR TITLE
Adding Simulation API

### PR DESCRIPTION
**Pull request by atatkin**
This pr adds a new `POST` route `api/simulate/:owner/:repo/:number`. This route requires a GitHub bearer token with access to read the the underlying pull request.

When called, policy-bot will re-evaluate the policy of the referenced pr and return a `simulated` result back to the caller. The result does NOT get written back to the pull request as a `STATUS`.

A request body can be included specifying a number of options which can modify how the simulation runs. These should be relatively easy to extend with additional options in the future.

Today this supports:
- Simulating new comments.
- Simulating new reviews.
- Simulating ignoring existing reviews and comments.
- Simulating a change of the `base` branch.

The route can also be used without any options to trigger policy-bot to re-evaluate the pull request as is.

 The returned simulated result does not currently contain all of the fields of the base `result` but it should be easy to extend as needed.

### CODE REVIEW

File: pull/simulated/context_test.go
Comment (bluekeyes) [inline pull/simulated/context_test.go]: The `pulltest` package provides an existing test implementation of `pull.Context` that you can use

File: pull/simulated/options.go
Comment (bluekeyes) [inline pull/simulated/options.go]: I think we should clarify and expand this field. Instead of combing both comments and reviews in a single field, I think having `IgnoreComments` and `IgnoreReviews` would be better. In the future, I could see adding an option to ignore commits or files and a generic `Ignore` field would be ambiguous.

I think these should also accept multiple entries so you could several users at once. It might even make sense to use the [`common.Actors`](https://github.com/palantir/policy-bot/blob/develop/policy/common/actor.go#L28) type here, which would allow ignoring items by organization, team, and permissions too.

If we do use the types from `common`, I think we should move this `simulated` package to be a subpackage of the `policy` package, instead of a subpackage of `pull`. It might make sense to do that anyway, since this is more of a decorator than a full implementation.

File: pull/simulated/options.go
Comment (bluekeyes) [inline pull/simulated/options.go]: More details below, but I think we'll need to take more information than just the identity of the author. Like with ignore rules, we should also accept multiple comments or reviews so that you can simulate policies with more complicated approval conditions.

A full simulation would also accept permission and membership details for these users, so that they don't have to actually exist in GitHub, but I think using real users is fine to start with.

File: pull/simulated/options.go
Comment (bluekeyes) [inline pull/simulated/options.go]: During evaluation, Policy Bot may list comments or reviews multiple times, relying on the `pull.Context` implementation to cache them after the first request. To avoid returning varying data, I think we'll want to use a fixed time here, either provided by the user, or set internally when the `simulated.Context` or the `Options` objects are created.

Taking the value from the user would allow things like testing `invalidate_on_push` by simulating comments that happened in the past.

File: pull/simulated/options.go
Comment (bluekeyes) [inline pull/simulated/options.go]: Individual rules in the policy can chose what strings count for approval, so we'll want to accept the comment content from the user instead of always using `:+1:`.

File: pull/simulated/options.go
Comment (bluekeyes) [inline pull/simulated/options.go]: Like with comments, we'll want to use a fix date. Reviews can optionally require certain text as well to count as an approval, so accepting a `Body` value might be nice.

File: server/handler/simulate.go
Comment (bluekeyes) [inline server/handler/simulate.go]: I think we're going to quickly run in to limitations trying to accept the simulation parameters as query parameters. Instead, I think we should make this a `POST` request and accept the simulation payload as a JSON body.

File: server/handler/simulate.go
Comment (bluekeyes) [inline server/handler/simulate.go]: I think we should avoid using `EvalContext` here and instead return the config and the simulation context directly. The `EvalContext` type is really designed to simplify posting statuses and doing other actions as part of evaluation that we don't need here and I don't want to accidentally start calling one of those functions.

File: server/handler/simulate_status.go
Comment (bluekeyes) [inline server/handler/simulate_status.go]: I think returning a `status` field with a different type based on whether the request was successful or an error will be confusing. I'd omit the status code here (it will already by in the HTTP response) and only include the `error` field (at which point, you could probably inline the whole thin in `writeAPIError`).

File: server/handler/simulate_status.go
Comment (bluekeyes) [inline server/handler/simulate_status.go]: For any situation where you are returning a 500 error, you could instead return the error directly (using `hatpear.Try` to wrap the handler) and rely on the [default baseapp error handler](https://github.com/palantir/go-baseapp/blob/develop/baseapp/error.go#L62-L65) to produce an appropriate response.

File: server/middleware/repo_auth.go
Comment (bluekeyes) [inline server/middleware/repo_auth.go]: All of these will turn in to generic "500 Internal Server Error" responses, which is not very helpful for authentication issues. I think we'll want to either write a response directly from this middleware or define a custom error that implements the [`StatusCode()`](https://github.com/palantir/go-baseapp/blob/develop/baseapp/error.go#L29-L31) method to take advantage of baseapp's ability to change response codes.

File: server/middleware/repo_auth.go
Comment (bluekeyes) [inline server/middleware/repo_auth.go]: We should handle 404 errors specially here since GitHub will return a 404 both if the repository does not exist and if the token doesn't have permissions.

File: server/middleware/repo_auth.go
Comment (bluekeyes) [inline server/middleware/repo_auth.go]: If the repository was transferred or renamed, the token can have access but the URL parameters may not match the API values. I think we can return `true` here without checking anything else?

File: server/middleware/repo_auth.go
Comment (bluekeyes) [inline server/middleware/repo_auth.go]: For application tokens, it's possible that an app will have permission to get repository metadata, but not have permission to read pull requests. Would it make sense to try reading the specific pull request instead, and if so, should that happen in the handler (where we have to read it anyway) instead of in middleware?

File: server/server.go
Comment (bluekeyes) [inline server/server.go]: As mentioned in the handler, I think this should be a POST request. I also don't think the `/status` suffix is necessary - to me, it is implied by the rest of the API.

File: server/server.go
Comment (bluekeyes) [inline server/server.go]: Can we move this up to L229 so all of the of the API routes are together?

**Review by bluekeyes — COMMENTED** [review]
Thanks for adding this and including simulation features beyond what we need for the initial use case. Overall, I think this implementation makes sense, but I had some suggestions for how the simulation options and the handler should work.

File: server/middleware/repo_auth.go
Comment (atatkin) [reply server/middleware/repo_auth.go]: Good idea, I updated the auth middleware to validate that the token can read the pr instead of the repo.

We could perform this lookup in the handler since we have to do that anyway, but it would require passing the bearer token through to the handler which probably means storing it in `context`.

I'm open to that if it's better to store the token in `context` vs having to lookup the same pull request twice. Though I do believe the github client is using caching, so perhaps that second lookup is not so bad?

Comment (bluekeyes) [reply server/middleware/repo_auth.go]: If we checked authentication in the handler, we'd probably skip the middleware completely. The handler would take care of everything using some helper function instead. The PR request should be cached, but it might still be nice to only do it once.

Comment (atatkin) [reply server/middleware/repo_auth.go]: oh I see so skip the middleware entirely and perform the token validation in the handler itself, that would work too since currently the simulate feature is a single route and I don't really see it growing to multiple routes anytime soon.

Comment (atatkin) [reply server/middleware/repo_auth.go]: Moved the token auth into the `simulate` handler and removed the now unneeded middleware.

File: policy/simulated/context_test.go
Comment (bluekeyes) [inline policy/simulated/context_test.go]: I think you only want to pass `message` in this assertion? Passing both `test` and `message` will try to use `test` as a format string.

This should also probably be `require.NoError` so that we don't run the `assert.Equals` when we already know the comments won't be correct.

File: policy/simulated/context_test.go
Comment (bluekeyes) [inline policy/simulated/context_test.go]: If we're going to extract the authors from `ExpectedComments` and otherwise not use any other fields, it might be easier to define the expectation as a `[]string` in the test struct.

File: policy/simulated/context_test.go
Comment (bluekeyes) [inline policy/simulated/context_test.go:164]: I recommend sorting this so that the test doesn't depend on the authors appearing in the right order.

File: policy/simulated/options.go
Comment (bluekeyes) [inline policy/simulated/options.go]: I think we can add `json` tags to the existing `common.Actors` struct to avoid have to redefine it here and have a conversion function.

File: policy/simulated/options.go
Comment (bluekeyes) [inline policy/simulated/options.go]: Looking at `IsActor`, it looks like it doesn't actually use the first `ctx` argument. If you want to simplify this and avoid passing a `context.Context` around, I think we could remove that argument.

File: policy/simulated/context.go
Comment (bluekeyes) [inline policy/simulated/context.go]: It feels odd to me to have most of the logic for the context implementation actually defined on the options type. Did you consider having `Options` only define functions for setting defaults and then defining the manipulation logic either on the `Context` type or as private functions in this file?

File: policy/simulated/options_test.go
Comment (bluekeyes) [inline policy/simulated/options_test.go]: You can use [`httptest.NewRequest`](https://pkg.go.dev/net/http/httptest#NewRequest) to avoid checking this error.

File: policy/simulated/options_test.go
Comment (bluekeyes) [inline policy/simulated/options_test.go]: `require.NoError` to avoid continuing the test when there is an error

File: server/apierror/error.go
Comment (bluekeyes) [inline server/apierror/error.go]: Could this be a function in the `handler` package or does it benefit from adding a new package?

If we do keep it in a new package, let's change the name to `Write` so that usage in other packages doesn't duplicate the "API error" part.

File: server/handler/simulate.go
Comment (bluekeyes) [inline server/handler/simulate.go]: This should be an internal error, I'm pretty sure `NewTokenClient` doesn't actually validate anything about the token and this will only happen if something fails creating the client instance (like when parsing the URL).

File: server/handler/simulate.go
Comment (bluekeyes) [inline server/handler/simulate.go]: If we're going to return a descriptive error to the client, there's usually no need to log it as well (unless there's extra internal information that is useful.) In particular, 4xx errors are usually noise in server logs because they indicate that the client did something wrong.

File: server/handler/simulate.go
Comment (bluekeyes) [inline server/handler/simulate.go]: nit:
```suggestion
		return errors.Wrap(err, "failed to get pull request")
```
for consistency with the other messages

File: server/handler/simulate.go
Comment (bluekeyes) [inline server/handler/simulate.go]: It might be worth returning an empty / nil result here, especially since picking a different base branch could easily lead to this error.

File: server/handler/simulate.go
Comment (bluekeyes) [inline server/handler/simulate.go]: `newSimulationResponse` also includes handling for when `result.Error` is not nil. I think we should defer to that and return the error in a "successful" simulation response, since that's how it would appear if Policy Bot encountered the same error in a real evaluation.

File: server/apierror/error.go
Comment (bluekeyes) [inline server/apierror/error.go]: It looks like all of the existing usage passes a single error message - I think we should force that via the API (right now, you could also provide no message). If we need multiple errors in the future, marshalling them as an `errors` field that is a list might be better.

File: policy/simulated/options.go
Comment (atatkin) [reply policy/simulated/options.go]: Yeah I considered that, but then `permissions` on `common.Actors` is defined as a `[]pull.Permission` which is a `uint8` and I thought it would be more friendly for the request body to reference the permissions by their string value and convert them into the `uint8` during the conversion function.

File: server/apierror/error.go
Comment (atatkin) [reply server/apierror/error.go]: Ah yes, it was intentionally written that way to allow an error response without a specific error message, but I ended up never doing that. I'll update it to require a single message.

File: server/apierror/error.go
Comment (atatkin) [reply server/apierror/error.go]: Yeah it can, it was in a separate package because it was being used by the `middleware` originally too, but now that we are performing the token check in the handler there is no need for it to be in a separate package. Will consolidate.

File: policy/simulated/context.go
Comment (atatkin) [reply policy/simulated/context.go]: Makes sense, updated so the logic was on `Context` instead of the options.

File: policy/simulated/options.go
Comment (atatkin) [reply policy/simulated/options.go]: Good idea, I looked into removing that but it's used a bunch of times by other packages and felt like it would bloat this pr, tracking: https://github.com/palantir/policy-bot/issues/727

File: policy/simulated/context_test.go
Comment (atatkin) [reply policy/simulated/context_test.go]: Nice catch, thanks!

File: policy/simulated/options.go
Comment (bluekeyes) [reply policy/simulated/options.go]: Ah, `pull.Permission` defines a custom `UnmarshalText` function so that you can use strings. While it's apparently missing the `yaml` tag, the strings work fine when used in YAML configuration and the `encoding/json` package will also use `UnmarshalText` when it is defined.

Comment (atatkin) [reply policy/simulated/options.go]: Oh nice, I missed that. I moved the `options` to use the built in structs and added the necessary json tags.

File: policy/common/actor.go
Comment (bluekeyes) [inline policy/common/actor.go]:
```suggestion
	Permissions []pull.Permission `yaml:"permissions" json:"permissions"`
```

Let's add the `yaml` tag too for consistency, it probably should have been there the whole time.

File: policy/common/actor.go
Comment (bluekeyes) [inline policy/common/actor.go]: If you want to ignore these when using JSON, let's add the `json:"-"` tag. Otherwise, they'll be unmarshalled, but use the field name for the keys.

File: policy/simulated/options.go
Comment (bluekeyes) [inline policy/simulated/options.go]: Sorry, I should have clarified in the last review: I think having custom simulation types for comments and reviews is the right idea. `common.Actors` is already used in public APIs (the configuration file), but the `pull` types are currently internal.

By adding `json` tags to `pull.Comment`/`pull.Review` and using them in the simulation API, we make them part of the public interface and will need to maintain compatibility going forward. They also have fields that aren't really relevant to the simulation right now (like ID, SHA, and teams for reviews).

Custom types here will let us decouple the internal handling of comments and reviews from what we accept in simulation requests so that we can evolve either part without affecting the other.

**Review by bluekeyes — COMMENTED** [review]
Thanks for all the fixes, I think this is almost ready to merge. Could you please also add a small section to the README about using the new API? I think it can go in the "Testing and Debugging Policies" section.

File: policy/simulated/options.go
Comment (atatkin) [reply policy/simulated/options.go]: Ack, i'll restore the custom types for the `pull.` types and keep the tags for the `common.Actors` type.

File: policy/simulated/options.go
Comment (bluekeyes) [inline policy/simulated/options.go]: Let's minimize the set of review fields we accept for now - we can always add more later, but taking away fields is harder. I think that means we should accept `CreatedAt,` `LastEditedAt`, `Author`, `Body`, and `State`. For everything else, let's either remove the field or set `json:"-"` so it is ignored when unmarshalling.

**Review by bluekeyes — COMMENTED** [review]
One more small suggestion on the `Review` type that I think we should address before merging, but then I think this is good to merge.

**Review by bluekeyes — APPROVED** [review]

**Pull request merged**
